ddl/column.go

```diff
@@ -712,6 +712,11 @@ func needChangeColumnData(oldCol, newCol *model.ColumnInfo) bool {
 		return needTruncationOrToggleSign()
 	}
 
+	if (types.IsTypeVarchar(oldCol.Tp) && newCol.Tp == mysql.TypeString) ||
+		(oldCol.Tp == mysql.TypeString && types.IsTypeVarchar(newCol.Tp)) {
```

Comment:
When the new collation is enabled, ALTER from char(N) to varchar(M) needs to reorganize the data for all of the primary key and secondary keys, because the key encoding is different.